base/twiceprecision.jl

```diff
@@ -0,0 +1,503 @@
+    Δ, Δfac = stop-start, 1
+    if ~isfinite(Δ)   # handle overflow for large endpoints
+        Δ, Δfac = stop/len - start/len, Int(len)
+    end
+    tmin = -(start/Δ)/Δfac            # interpolation t such that return value is 0
+    imin = round(Int, tmin*(len-1)+1)
+    if 1 < imin < len
+        # The smallest-magnitude element is in the interior
+        t = (imin-1)/(len-1)
+        ref = T((1-t)*start + t*stop)
+        step = imin-1 < len-imin ? (ref-start)/(imin-1) : (stop-ref)/(len-imin)
+    elseif imin <= 1
+        imin = 1
+        ref = start
+        step = (Δ/(len-1))*Δfac
+    else
+        imin = Int(len)
+        ref = stop
+        step = (Δ/(len-1))*Δfac
+    end
+    if len == 2 && !isfinite(step)
+        # For very large endpoints where step overflows, exploit the
+        # split-representation to handle the overflow
+        return StepRangeLen(TwicePrecision(start, zero(T)),
+                            TwicePrecision(-start, stop), 2)
+    end
+    # 2x calculations to get high precision endpoint matching while also
+    # preventing overflow in ref_hi+(i-offset)*step_hi
+    t, k = prevfloat(realmax(T)), max(imin-1, len-imin)
+    step_hi_pre = clamp(step, max(-(t+ref)/k, (-t+ref)/k), min((t-ref)/k, (t+ref)/k))
+    nb = nbitslen(len, imin)
+    step_hi = truncbits(step_hi_pre, nb)
+    x1_hi, x1_lo = add2((1-imin)*step_hi, ref)
+    x2_hi, x2_lo = add2((len-imin)*step_hi, ref)
+    a, b = (start - x1_hi) - x1_lo, (stop - x2_hi) - x2_lo
+    step_lo = (b - a)/(len - 1)
+    ref_lo = abs(ref) < eps(max(abs(start), abs(stop))) ? zero(T) : a - (1 - imin)*step_lo
+    StepRangeLen(TwicePrecision(ref, ref_lo), TwicePrecision(step_hi, step_lo), Int(len), imin)
+end
+
+# linspace for rational numbers, start = start_n/den, stop = stop_n/den
+# Note this returns a StepRangeLen
+function linspace{T}(::Type{T}, start_n::Integer, stop_n::Integer, len::Integer, den::Integer)
+    len < 2 && return _linspace1(T, start_n/den, stop_n/den, len)
+    start_n == stop_n && return StepRangeLen(TwicePrecision{T}((start_n, den)), zero(TwicePrecision{T}), len)
+    tmin = -start_n/(Float64(stop_n) - Float64(start_n))
+    imin = round(Int, tmin*(len-1)+1)
+    imin = clamp(imin, 1, Int(len))
+    # Compute (1-t)*a and t*b separately in 2x precision (itp = interpolant)...
+    dent = (den, len-1)  # represent products as a tuple to eliminate risk of overflow
+    start_itp = proddiv(T, (len-imin, start_n), dent)
+    stop_itp = proddiv(T, (imin-1, stop_n), dent)
+    # ...and then combine them to make ref
+    ref = start_itp + stop_itp
+    # Compute step to 2x precision without risking overflow...
+    rend = proddiv(T, (stop_n,), dent)
+    rbeg = proddiv(T, (-start_n,), dent)
+    step = twiceprecision(rbeg + rend, nbitslen(len, imin)) # ...and truncate hi-bits as needed
+    StepRangeLen(ref, step, Int(len), imin)
+end
+
+# For len < 2
+function _linspace1{T}(::Type{T}, start, stop, len)
```

Comment:
it only ever gets called this way, but `len::Integer` just to re-affirm that locally?